crates/sdk/src/error.rs

```diff
@@ -0,0 +1,96 @@
+use std::sync::Arc;
+
+use thiserror::Error;
+
+#[derive(Error, Debug, Clone)]
+#[non_exhaustive]
+pub enum Error {
+    #[error("Disconnected normally following a call to `DbContext::disconnect`")]
+    Disconnected,
+
+    #[error("Already disconnected in call to `DbContext::disconnect`")]
+    AlreadyDisconnected,
+
+    #[error("Host returned error when processing subscription query: {error}")]
+    SubscriptionError { error: String },
+
+    #[error("Subscription has already ended")]
+    AlreadyEnded,
+
+    #[error("Unsubscribe already called on subscription")]
+    AlreadyUnsubscribed,
+
+    #[error("Error in WebSocket connection: {0}")]
+    Ws(#[from] crate::websocket::WsError),
+
+    #[error(transparent)]
+    Internal(#[from] InternalError),
+}
+
+#[derive(Debug, Clone)]
+pub struct InternalError {
+    message: String,
+    cause: Option<Arc<dyn std::error::Error + Send + Sync>>,
+}
+
+impl InternalError {
+    pub(crate) fn new(message: impl Into<String>) -> Self {
+        Self {
+            message: message.into(),
+            cause: None,
+        }
+    }
+
+    #[doc(hidden)]
+    /// Called by codegen. Not part of this library's stable API.
+    pub fn with_cause(self, cause: impl std::error::Error + Send + Sync + 'static) -> Self {
+        Self {
+            cause: Some(Arc::new(cause)),
+            ..self
+        }
+    }
+
+    #[doc(hidden)]
+    /// Called by codegen. Not part of this library's stable API.
+    pub fn failed_parse(ty: &'static str, container: &'static str) -> Self {
+        Self::new(format!("Failed to parse {ty} from {container}"))
```

Comment:
Does it make sense to add a suggestion in the error message to regen the client API? Or is this usually due to actual bugs.

Reply:
That is an excellent idea, and I will do so.